### PR TITLE
Strictly enforce ASN.1 DER where possible

### PR DESCRIPTION
**Pull request by DemiMarie**
## Description

Please write a few sentences describing the overall goals of the pull request's commits.

This makes the ASN.1 parser much stricter.  It rejects negative integers everywhere, and enforces DER almost everywhere, with at least 4 exceptions:

1. Boolean TRUE values are allowed to be any positive 8-bit number, not just 0xFF.
2. Omitting the `Z` in a `UTCTime` or `GeneralizedTime` is allowed.
3. Ordering of entries in a `SET` or `SET OF` is not checked.
4. Requiring default values to be omitted must be done in higher-level code.  I only removed the checks for trailers in RSASSA-PSS algorithm identifiers.

Exceptions 1 and 2 exist because many of the X.509 tests rely on them.  Exception 3 exists because I have not implemented checks that would exclude it.  4 is a cross-cutting concern as long as the ASN.1 parsers in Mbed TLS are written by hand instead of being machine-generated.

The most challenging part of this was getting the test suite to pass.  Some of the needed changes fix bugs in the testsuite, and others alter tests that explicitly checked for behavior on invalid DER.

## Gatekeeper checklist
 - [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

## Notes for the submitter

Please refer to the [contributing guidelines](../CONTRIBUTING.md), especially the
checklist for PR contributors.

Fixes: #3329
Fixes: #4070

### CODE REVIEW

**Comment by davidhorstmann-arm** [comment]
Thanks for your contribution! We use Developer Certificate of Origin (see [the contributing guide](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md#license-and-copyright)), please add a `Signed-off-by:` line to the commit messages. You can do this easily with:
```
git rebase --signoff <start commit>~
```

File: library/asn1parse.c [hunk 1]
```diff
@@ -219,11 +244,23 @@ int mbedtls_asn1_get_bitstring( unsigned char **p, const unsigned char *end,
...
-    bs->unused_bits = **p;
-    if( bs->unused_bits > 7 )
-        return( MBEDTLS_ERR_ASN1_INVALID_LENGTH );
-    (*p)++;
+    /* Get number of unused bits */
+    if ( ( bs->unused_bits = *(*p)++ ) )
+    {
+        unsigned char last;
+
+	/*
```
Comment (tom-cosgrove-arm) [inline library/asn1parse.c:252]: Please don't use tab characters - this is enforced by the CI, so the PR won't pass when any tabs are used for indenting

**Comment by tom-cosgrove-arm** [comment]
The Travis log has

```
test_suite_psa_crypto ............................................. suites/test_suite_psa_crypto.function:96:5: runtime error: addition of unsigned offset to 0x61d0000009d4 overflowed to 0x61d0000008d3
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior suites/test_suite_psa_crypto.function:96:5 in 
FAIL
```

and the Windows build fails on the Internal CI

```
       "C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj" (Rebuild target) (306) ->

       (ClCompile target) -> 

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2146: syntax error : missing ')' before identifier 'actual_length' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2081: 'ssize_t' : name in formal parameter list illegal [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2061: syntax error : identifier 'actual_length' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ';' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ',' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

         C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ')' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]

    164 Warning(s)
    6 Error(s)
```

**Comment by DemiMarie** [comment]
> The Travis log has
> 
> ```
> test_suite_psa_crypto ............................................. suites/test_suite_psa_crypto.function:96:5: runtime error: addition of unsigned offset to 0x61d0000009d4 overflowed to 0x61d0000008d3
> SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior suites/test_suite_psa_crypto.function:96:5 in 
> FAIL
> ```

Bug in my testsuite changes.

> and the Windows build fails on the Internal CI
> 
> ```
>        "C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj" (Rebuild target) (306) ->
> 
>        (ClCompile target) -> 
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2146: syntax error : missing ')' before identifier 'actual_length' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2081: 'ssize_t' : name in formal parameter list illegal [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2061: syntax error : identifier 'actual_length' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ';' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ',' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>          C:/builds/workspace/mbed-tls-pr-merge_PR-6659-merge/worktrees/tmpa5h2t0y0/tests/suites/test_suite_asn1parse.function(125): error C2059: syntax error : ')' [C:\builds\workspace\mbed-tls-pr-merge_PR-6659-merge\worktrees\tmpa5h2t0y0\cmake_solution\tests\test_suite_asn1parse.vcxproj]
> 
>     164 Warning(s)
>     6 Error(s)
> ```

Windows does not have `ssize_t`.  Also fixed.

**Comment by DemiMarie** [comment]
Would it be helpful if I split this into several smaller PRs to ease review?

**Comment by DemiMarie** [comment]
Closing as this PR is too big to be reviewable.  I will submit the individual parts separately.

**Pull request closed without merging**